### PR TITLE
Jitter Buffer implementation for rtpengine

### PR DESCRIPTION
**Pull request by balajeesv**
Hi,
  Please review Jitter buffer implementation in rtpengine. Discussions regarding jitter buffer implementation can be found [here](https://github.com/sipwise/rtpengine/issues/803).

The algorithm is mentioned in the [gist](https://gist.github.com/balajeesv/1a0cec2a92ed6eb5299e56bd1691457b) and the timer is implemented with existing timer thread pool instead of ptime proposed. 

The existing timer is enhanced to support jitter buffer by introducing a few changes like insert into the queue with sorting in descending order (sorted in descending order for faster insertion) and read from the tail. 

Tested this code with basic buffering, packets drops, duplicates, repacketization, and transcoding.

Thanks,
Balajee

### CODE REVIEW

**Comment by balajeesv** [comment]
With new commit sender clock drift handling and few other minor fixes are done.Please check.

**Comment by balajee-plivo** [comment]
@rfuchs  Did you get chance to review this PR

**Comment by balajeesv** [comment]
@rfuchs Did you get a chance to review this PR?

**Comment by rfuchs** [comment]
No not yet, sorry for the delay, I'll try to get to it one of these days

File: daemon/call.c [hunk 1]
```diff
@@ -899,11 +899,13 @@ struct packet_stream *__packet_stream_new(struct call *call) {
...
 	mutex_init(&stream->in_lock);
 	mutex_init(&stream->out_lock);
+	mutex_init(&stream->jb.lock);
 	stream->call = call;
 	atomic64_set_na(&stream->last_packet, rtpe_now.tv_sec);
 	stream->rtp_stats = g_hash_table_new_full(g_int_hash, g_int_equal, NULL, __rtp_stats_free);
 	recording_init_stream(stream);
-	stream->send_timer = send_timer_new(stream);
+	stream->send_timer = send_timer_new(stream, 0);
+	stream->buffer_timer = send_timer_new(stream, 1);
```
Comment (rfuchs) [inline daemon/call.c]: These should be either different functions (send_timer_new and jitter_buffer_new or something like that), or only use a single timer object.

File: daemon/media_socket.c [hunk 2]
```diff
@@ -1677,7 +1678,7 @@ static int do_rtcp(struct packet_handler_ctx *phc) {
 int media_socket_dequeue(struct media_packet *mp, struct packet_stream *sink) {
 	struct codec_packet *p;
 	while ((p = g_queue_pop_head(&mp->packets_out)))
-		send_timer_push(sink->send_timer, p);
+		send_timer_push((p->packet && p->packet->buffered)?sink->buffer_timer:sink->send_timer, p);
```
Comment (rfuchs) [inline daemon/media_socket.c]: Should be able to eliminate this case distinction.

File: daemon/jitter_buffer.c [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+#define CONT_INCORRECT_BUFFERING 0x14
+
+
+static struct jitter_buffer_config *jb_config; 
+
+struct jitter_buffer_config {
+	int    min_jb_len;
+	int    max_jb_len;
+	int    enable_jb;
+	int    src_clockrate;
```
Comment (rfuchs) [inline daemon/jitter_buffer.c]: I'm a bit confused about this parameter as I don't really see it being used, other than as an enable switch for the clock drift calculation.

File: daemon/jitter_buffer.c [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+		(long long) (ts_diff)* 1000000 / clockrate;
+	struct timeval to_send = jb->first_send;
+	timeval_add_usec(&to_send, ts_diff_us);
+	long long time_diff = timeval_diff(&rtpe_now, &to_send);
+
+	jb->clock_drift_val = time_diff/seq_diff;
+	jb->clock_drift_enable = 0;
+	jb->cont_buff_err = 0;
+}
+
```
Comment (rfuchs) [inline daemon/jitter_buffer.c]: Can you explain the algorithm used here? AFAICT this is a single-shot calculation that is done when an overflow or underflow occurs?

**Comment by balajeesv** [comment]
> Ok, so from a quick first glace, I see that this is implemented as an incoming packet buffer, unlike the send_timer mechanism that already exists, which is an outgoing packet buffer. Originally I was under the impression that this would be another outgoing packet buffer, which is why I suggested using the send_timer.
> 
> This makes me think that a more generic version of send_timer would be more appropriate in order to eliminate some unnecessary case distinctions. Either completely unlinked from the send_timer as another instance of a timerthread (which would require separate dedicated threads), or within the send_timer but using a handler function in codec_packet (which then should probably be renamed).
> 
> I can probably handle this kind of refactoring if you prefer.

Incoming packet buffer used because of transcoding and re-packetization would get benefitted from the incoming buffer.
I would be happy to get this refactored to a more generic version of send_timer :)

**Comment by balajee-plivo** [comment]
@rfuchs Did you get chance to look into responses?

**Comment by rfuchs** [comment]
Yes, thank you, I plan on getting on this shortly.

**Comment by rfuchs** [comment]
Just a quick update here to let you know that I'm almost through with refactoring and integrating your code.

**Comment by rfuchs** [comment]
I've pushed my refactored version of this into the branch `rfuchs/jb`. If you can, please try it and see if/how it works for you.

**Comment by balajee-plivo** [comment]
Sure. Will test and update you.

**Comment by balajeesv** [comment]
Sure. I will test and update you.

**Pull request closed without merging**